pkg/upgrade/upgrader/jiva_upgrade.go

```diff
@@ -281,47 +282,61 @@ func getPVCDeploymentsNamespace(
 	return ns, nil
 }
 
-func validateSync(ctrlLabel, namespace string) error {
-	klog.Infof("Verifying replica sync")
-	quorum := false
+func getReplicationFactor(ctrlLabel, namespace string) (int, error) {
 	ctrlList, err := podClient.WithNamespace(namespace).List(
 		metav1.ListOptions{
 			LabelSelector: ctrlLabel,
 		})
 	if err != nil {
-		return err
+		return 0, err
 	}
 	if len(ctrlList.Items) == 0 {
-		return errors.Errorf("no deployments found for %s in %s", ctrlLabel, namespace)
+		return 0, errors.Errorf("no deployments found for %s in %s", ctrlLabel, namespace)
 	}
 	ctrlPod := ctrlList.Items[0]
+	return strconv.Atoi(ctrlPod.Spec.Containers[0].Env[0].Value)
+}
+
+func getAPIURL(svcLabel, namespace string) (string, error) {
+	svcList, err := serviceClient.WithNamespace(namespace).List(
+		metav1.ListOptions{
+			LabelSelector: svcLabel,
+		})
+	if err != nil {
+		return "", err
+	}
+	if len(svcList.Items) == 0 {
+		return "", errors.Errorf("no service found for %s in %s", svcLabel, namespace)
+	}
+	targetIP := svcList.Items[0].Spec.ClusterIP
+	apiURL := "http://" + targetIP + ":9501/v1/replicas"
+	return apiURL, nil
+}
+
+func validateSync(pvLabel, namespace string) error {
+	klog.Infof("Verifying replica sync")
+	ctrlLabel := "openebs.io/controller=jiva-controller," + pvLabel
+	svcLabel := "openebs.io/controller-service=jiva-controller-svc," + pvLabel
+	quorum := false
 	syncedReplicas := 0
-	replicationFactor, err := strconv.Atoi(ctrlPod.Spec.Containers[0].Env[0].Value)
+	replicationFactor, err := getReplicationFactor(ctrlLabel, namespace)
+	if err != nil {
+		return err
+	}
+	apiURL, err := getAPIURL(svcLabel, namespace)
 	if err != nil {
 		return err
 	}
 	for syncedReplicas != replicationFactor {
 		syncedReplicas = 0
-		out, err := podClient.WithNamespace(ctrlPod.Namespace).
-			Exec(
-				ctrlPod.Name,
-				&corev1.PodExecOptions{
-					Command: []string{
-						"/bin/bash",
-						"-c",
-						"curl http://localhost:9501/v1/replicas",
-					},
-					Container: ctrlPod.Spec.Containers[0].Name,
-					Stdin:     false,
-					Stdout:    true,
-					Stderr:    true,
-				},
-			)
+		httpClient := &http.Client{Timeout: 2 * time.Second}
+		resp, err := httpClient.Get(apiURL)
 		if err != nil {
 			return err
 		}
 		replicas := jivaClient.ReplicaCollection{}
-		err = json.Unmarshal([]byte(out.Stdout), &replicas)
+		defer resp.Body.Close()
```

Comment:
move defer up just after the error check (idiomatic go)

Reply:
Done